pam_xdg.8
.\"@ pam_xdg - manage XDG Base Directories (runtime dir life time, environ).
.\"
.\" Copyright (c) 2022 - 2024 Steffen Nurpmeso <steffen@sdaoden.eu>.
.\" SPDX-License-Identifier: ISC
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.
.Dd Sep 3, 2026
.Dt PAM_XDG 8
.Os
.
.
.Sh NAME
.Nm pam_xdg.so
.Nd PAM module that manages XDG Base Directories
.
.
.Sh SYNOPSIS
.
.Nm
.Op Ar runtime
.Op Ar notroot
.Op Ar track_sessions Op Ar per_user_lock
.Op Ar per_user_cgroup=/path/to/BASE
.
.
.Sh DESCRIPTION
.
.Nm
is a PAM module that manages creation of the
.Ev XDG_RUNTIME_DIR
directory, as well as injection of environment variables denoting all
directories specified by the
.Lk https://specifications.\:freedesktop.\:org/basedir-\:\
spec/\:basedir-\:spec-\:latest.html \
"XDG Base Directory Specification 0.8" ,
into user sessions.
When linked into the PAM session system the runtime directory will be
created once a user creates his or her first login session.
On Linux it can optionally place sessions in dedicated per-user cgroups.
.
.Pp
To make use of this module, place for example the following in the
.Ql session
part of the control file of desire under
.Pa /etc/pam.d ,
on Linux it may be
.Pa /etc/pam.d/common-session
if that exists, on BSD's the files
.Pa /etc/pam.d/system
as well as
.Pa /etc/pam.d/login ,
.Pa /etc/pam.d/sshd
and
.Pa /etc/pam.d/su
may be desirable, adjusting paths as necessary:
.
.Bd -literal -offset indent
session optional pam_xdg.so notroot track_sessions
.Ed
.
.Pp
If
.Ar notroot
was given the module will bypass itself for root account logins and
perform no actions for root, except removing any possibly existing XDG
environment variable.
.
.Pp
Unless
.Ar runtime
was given all XDG related environment variables will be created in all
user sessions with their default or computed values, otherwise only
.Ev XDG_RUNTIME_DIR .
.
.Pp
.Ar track_sessions
will enable session tracking: once the last session ends, the user's
.Ev XDG_RUNTIME_DIR
will be recursively removed; on high-load servers then setting
.Ar per_user_lock
will reduce lock file lock contention.
.
.Pp
.Ar per_user_cgroup=/path/to/BASE
can be set on Linux to a directory within a mounted
.Ql cgroup2
filesystem, one element of which,
.Ar BASE ,
will be automatically created as necessary.
For example,
.Ql per_user_cgroup=/sys/fs/cgroup/_user
will try to create the
.Ql _user
directory when yet non-existent;
within this directory, per-user directories will be created, like
.Ql /sys/fs/cgroup/_user/1000 ,
and the entire user session be moved to that cgroup.
(Errors in doing any of this is logged but ignored.)
.
.
.Sh "SEE ALSO"
.
.Xr pam 3 ,
.Xr pam.conf 5
.
.
.Sh AUTHORS
.
.An "Steffen Nurpmeso" Aq steffen@sdaoden.eu .
.
.
.Sh CAVEATS
.
On Unix systems any
.Dq daemonized
program or script is reparented to the program running with PID 1,
most likely leaving the PAM user session without PAM recognizing this.
Yet careless such code may hold or expect availability of resources of
the session it just left, truly performing cleanup when sessions end
seems thus unwise.
Since so many PAM modules do support session tracking and cleanup
.Nm
readded optional support for this.
On Linux the
.Ar per_user_cgroup
mechanism accesses an hierarchical grouping that cannot be escaped.
.
.\" s-ts-mode


pam_xdg.c
/*@ pam_xdg - manage XDG Base Directories (runtime dir life time, environment).
 *@ See pam_xdg.8 for more.
 *@ - According to XDG Base Directory Specification, v0.8.
 *@ - Supports libpam (Linux) and OpenPAM.
 *@ - Requires C preprocessor with __VA_ARGS__ support!
 *@ - Uses "rm -rf" to drop per-user directories. XXX Unroll this?  nftw?
 *@   Problems are (also dependent on operating system)
 *@   E[MN]FILE failures, ordering issues, mode changes, mounts, subvolumes..
 *@
 *@ Thanks:
 *@	Jan Beich	(jbeich at FreeBSD dot org)
 *@	Andre Albsmeier	(mail at fbsd2 dot e4m dot org)
 *
 * Copyright (c) 2021 - 2026 Steffen Nurpmeso <steffen@sdaoden.eu>.
 * SPDX-License-Identifier: ISC
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#ifndef XDG_CONFIG_DIR
# define XDG_CONFIG_DIR /etc
#endif
#ifndef XDG_DATA_DIR_LOCAL
# define XDG_DATA_DIR_LOCAL /usr/local
#endif
#ifndef XDG_RUNTIME_DIR_OUTER
# define XDG_RUNTIME_DIR_OUTER /run
#endif

/* For these a leading \1 is replaced with struct passwd::pw_dir.
 * Be aware we use a stack buffer for storage */
#define a_XDG_CACHE_HOME_DEF "\1/.cache"
#define a_XDG_CONFIG_HOME_DEF "\1/.config"
#define a_XDG_DATA_HOME_DEF "\1/.local/share"
#define a_XDG_STATE_HOME_DEF "\1/.local/state"

#define a_XDG_CONFIG_DIRS_DEF a_STRING(XDG_CONFIG_DIR) "/xdg"
#define a_XDG_DATA_DIRS_DEF a_STRING(XDG_DATA_DIR_LOCAL) "/share:/usr/share"

/* We create the outer directories as necessary (stack buffer storage!).
 * This only holds for last component of _OUTER, though. */
#define a_RUNTIME_DIR_OUTER a_STRING(XDG_RUNTIME_DIR_OUTER)
#define a_RUNTIME_DIR_OUTER_MODE 0755
#define a_RUNTIME_DIR_BASE "user"
#define a_RUNTIME_DIR_BASE_MODE 0755 /* 0711? */

/* Note: we manage these relative to the per-user directory!
 * a_LOCK_FILE is only used without "per_user_lock" */
#define a_LOCK_FILE "../." a_XDG ".lck"
#define a_LOCK_TRIES 10

#define a_DAT_FILE "." a_XDG ".dat"

/* CGROUP support 0/1 stuff (on Linux), only used with per_user_cgroup=/X/Y/DIR_BASE/..user.. */
#define a_CGROUP 1
#define a_CGROUP_DIR_BASE_MODE 0755
#define a_CGROUP_PROCS_FILE "cgroup.procs"

/* >8 -- 8< */

/*
#define _POSIX_C_SOURCE 200809L
#define _ATFILE_SOURCE
*/
#define _GNU_SOURCE /* Always the same mess */

#include <sys/stat.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <pwd.h>
#include <stdint.h> /* xxx not, actually!?! */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <security/pam_appl.h>
#include <security/pam_modules.h>
#ifdef OPENPAM_VERSION
# include <security/openpam.h>
#elif !defined __solaris__ && !defined __sun
# include <security/pam_ext.h>
#endif

#ifdef OPENPAM_VERSION
#else
# include <syslog.h>
#endif

/* Who are we? */
#define a_XDG "pam_xdg"

/* _XOPEN_PATH_MAX POSIX 2008/Cor 1-2013 */
#ifndef PATH_MAX
# define PATH_MAX 1024
#endif

/* */
#ifdef O_SEARCH
# define a_O_SEARCH O_SEARCH
#elif defined O_PATH
# define a_O_SEARCH O_PATH
#else
  /* Well, hardly, but not in practice so do not #error out */
# define a_O_SEARCH 0
#endif

/* */
#define a_STRING(X) a__STRING(X)
#define a__STRING(X) #X

/* libpam / OpenPAM compat */
#ifdef OPENPAM_VERSION
# define a_LOG(HDL, LVL, ...) ((void)HDL, openpam_log(LVL, __VA_ARGS__))
# define a_LOG_ERR PAM_LOG_ERROR
# define a_LOG_NOTICE PAM_LOG_NOTICE
#elif !defined __solaris__ && !defined __sun
# define a_LOG(HDL, LVL, ...) pam_syslog(HDL, LVL, __VA_ARGS__)
#else
# define a_LOG(HDL, LVL, ...) syslog(LVL, __VA_ARGS__)
# define a_GET_ITEM_ARG_CAST(X) (void**)X
#endif

#ifndef a_LOG_ERR
# define a_LOG_ERR LOG_ERR
# define a_LOG_NOTICE LOG_NOTICE
#endif
#ifndef a_GET_ITEM_ARG_CAST
# define a_GET_ITEM_ARG_CAST(X) (void const**)X
#endif

#if a_CGROUP && (defined __linux__ || defined __linux)
# undef a_CGROUP
# define a_CGROUP 1
#else
# undef a_CGROUP
# define a_CGROUP 0
#endif

/* Just put it all in one big fun, use two exec paths */
static int a_xdg(int isopen, pam_handle_t *pamh, int flags, int argc, char const **argv);

static int
a_xdg(int isopen, pam_handle_t *pamh, int flags, int argc, char const **argv){
	enum a_flags{
		a_NONE,
		/* Options */
		a_RUNTIME = 1u<<0,
		a_NOTROOT = 1u<<1,
		a_SESSIONS = 1u<<2,
		a_USER_LOCK = 1u<<3,
#if a_CGROUP
		a_USER_CGROUP = 1u<<4,
#endif

		a_SKIP_XDG = 1u<<15, /* We shall not act */

		a__SAVED_MASK = (a_SKIP_XDG << 1) - 1, /* Bits to restore on !isopen */

		/* Flags */
		a_MPV = 1u<<30 /* Multi-Purpose-Vehicle */
	};

	struct a_dirtree{
		char const *name;
		int mode;
	};

	static struct a_dirtree const a_dirtree[] = {
		{a_RUNTIME_DIR_OUTER, a_RUNTIME_DIR_OUTER_MODE},
		{a_RUNTIME_DIR_BASE, a_RUNTIME_DIR_BASE_MODE},
		{NULL, 0} /* XXX -> nelem/item/countof */
	};
	static int f_saved;

	char uidbuf[sizeof "../.18446744073709551615"],
#if a_CGROUP
		pidbuf[sizeof "18446744073709551615\n"],
#endif
		xbuf[((sizeof("XDG_RUNTIME_DIR=") + sizeof(a_RUNTIME_DIR_OUTER) +
			sizeof(a_RUNTIME_DIR_BASE) + sizeof("../.18446744073709551615")) |
			(sizeof("XDG_CONFIG_DIRS=") + PATH_MAX)
		) +1];
	struct a_dirtree dt_user;
	struct a_dirtree const *dtp;
	struct passwd *pwp;
	int cwdfd, cntrlfd, datfd, f, res, uidbuflen;
	char const *user, *emsg
#if a_CGROUP
		, *cgroup, *cgroup_end
#endif
		;

	user = "<unset>";
	cwdfd = AT_FDCWD;
	datfd = cntrlfd = -1;

	/* {{{ Command line */
	if(isopen){
		f = a_NONE;

		for(; argc > 0; ++argv, --argc){
			if(!strcmp(argv[0], "runtime"))
				f |= a_RUNTIME;
			else if(!strcmp(argv[0], "notroot"))
				f |= a_NOTROOT;
			else if(!strcmp(argv[0], "track_sessions"))
				f |= a_SESSIONS;
			else if(!strcmp(argv[0], "per_user_lock"))
				f |= a_USER_LOCK;
#if a_CGROUP
			else if(!strncmp(argv[0], "per_user_cgroup=", sizeof("per_user_cgroup=") -1)){
				size_t i;

				f |= a_USER_CGROUP;
				cgroup = &argv[0][sizeof("per_user_cgroup=") -1];
				i = strlen(cgroup);
				cgroup_end = &cgroup[i] - 1;

				if(i == 0 || *cgroup != '/'){
					emsg = "per_user_cgroup= empty or not absolute path";
					f ^= a_USER_CGROUP;
				}else if(i >= sizeof(xbuf) -1 - sizeof(uidbuf) -1 - 1 - sizeof(a_CGROUP_PROCS_FILE) -1){
					emsg = "per_user_cgroup= argument too long";
					f ^= a_USER_CGROUP;
				}else if(*cgroup_end == '/'){
					emsg = "per_user_cgroup= must not end with directory separator";
					f ^= a_USER_CGROUP;
				}

				if(!(f & a_USER_CGROUP)){
					errno = EINVAL;
					goto jerr;
				}
			}
#endif /* a_CGROUP */
			else if(!(flags & PAM_SILENT)){
				emsg = "command line";
				errno = EINVAL;
				goto jerr;
			}
		}

		if((f & (a_SESSIONS | a_USER_LOCK)) == a_USER_LOCK)
			a_LOG(pamh, a_LOG_NOTICE, a_XDG ": \"per_user_lock\" requires \"track_sessions\"");
	}else{
		f = f_saved;
		if(f & a_SKIP_XDG)
			goto jok;
	} /* }}} */

	/* We need the user we go for */
	if((res = pam_get_item(pamh, PAM_USER, a_GET_ITEM_ARG_CAST(&user))) != PAM_SUCCESS ||
			user == NULL || *user == '\0'){
		user = "<lookup failed>";
		emsg = "cannot query PAM_USER name";
		goto jepam;
	}

	/* No PAM failure, no PAM_USER_UNKNOWN here: we are no authentificator! */
	if((pwp = getpwnam(user)) == NULL){
		emsg = "host does not know about user";
		errno = EINVAL;
		goto jerr;
	}

	if((f & a_NOTROOT) && pwp->pw_uid == 0){
		/* We may switch within a login (eg "su"); however, clearing XDG environment variables is impossible:
		 *   pam_putenv: delete non-existent entry; XDG_RUNTIME_DIR
		 *   pam_xdg: user root: pam_putenv(): Bad item passed to pam_*_item()
		 * This is true even if we set them first (in effect).
		 * So we *would* set these but do not because of "notroot", therefore we *should* clear them it seems.
		 * (https://forums.freebsd.org/threads/using-sysutils-pam_xdg-but-xdg_xxx-vars-are-not-updating-
		 *	upon-su-and-maybe-in-other-situations-too.87741)
		 * We therefore mess directly with unsetenv(3) */
		if(isopen){
			static char const a_dirs[] = {
				"XDG_RUNTIME_DIR\0"
				"XDG_CACHE_HOME\0" "XDG_CONFIG_DIRS\0" "XDG_CONFIG_HOME\0"
				"XDG_DATA_DIRS\0" "XDG_DATA_HOME\0" "XDG_STATE_HOME\0"
				""
			};

			for(emsg = a_dirs; *emsg != '\0';){
				if((res = unsetenv(emsg)) != 0)
					a_LOG(pamh, a_LOG_ERR, a_XDG ": notroot: cannot unsetenv(): %s: %s\n",
						emsg, strerror(errno));
				emsg += strlen(emsg) +1;
			}
		}

		f |= a_SKIP_XDG;
		goto jok;
	}

	/* Our lockfile and per-user directory name */
	uidbuflen = snprintf(uidbuf, sizeof(uidbuf), "../.%lu", (unsigned long)pwp->pw_uid) - 3; /* xxx error?? */

	dt_user.name = &uidbuf[4];
	dt_user.mode = 0700; /* XDG implied */

	/* {{{ Handle tree, go to user runtime.  On *BSD outermost may not exist! */
	for(/*f &= ~a_MPV,*/ dtp = a_dirtree;;){
		int e;
		gid_t oegid;
		mode_t oumask;

		if((res = openat(cwdfd, dtp->name, (a_O_SEARCH | O_DIRECTORY | O_NOFOLLOW))) != -1){
			if(cwdfd != AT_FDCWD)
				close(cwdfd); /* XXX error hdl */
			cwdfd = res;

			if(dtp == &dt_user)
				break;
			else if((++dtp)->name == NULL)
				dtp = &dt_user;
			f &= ~a_MPV;
			continue;
		}

		if(!isopen)
			/* XXX Entire directory tree disappeared while sessions were open!
			 * XXX Silently out!?! */
			goto jok;

		/* We try creating the directories once as necessary */
		if((f & a_MPV) || errno != ENOENT){
			emsg = "cannot obtain chdir(2) descriptor (within) tree "
					a_RUNTIME_DIR_OUTER "/" a_RUNTIME_DIR_BASE;
			goto jerr;
		}
		f |= a_MPV;

		/* (On at least Linux, in containers/namespaces and whatever weird execution environments,
		 * we need to do this to get it graceful (i think of it as a PAM bug; undocumented, anyway)) */
		oumask = umask(0000);
		oegid = getegid();
		setegid(0);
			res = mkdirat(cwdfd, dtp->name, dtp->mode);
			e = (res == -1) ? errno : 0;
		setegid(oegid);
		umask(oumask);

		if(res == -1){
			if(e != EEXIST){
				emsg = "cannot create directory (within) tree "
						a_RUNTIME_DIR_OUTER "/" a_RUNTIME_DIR_BASE;
				goto jerr;
			}
		}else if(cwdfd == AT_FDCWD)
			a_LOG(pamh, a_LOG_NOTICE, a_XDG ": " a_RUNTIME_DIR_OUTER " did not exist, but should be "
				"(a mount point of) volatile storage!");
		/* Simply chown it! */
		else if(dtp == &dt_user &&
				fchownat(cwdfd, &uidbuf[4], pwp->pw_uid, pwp->pw_gid, AT_SYMLINK_NOFOLLOW) == -1){
			emsg = "cannot chown(2) per user XDG_RUNTIME_DIR";
			goto jerr;
		}
	} /* }}} */

	/* {{{ When opening, put environment, handle cgroup.  Ignore (but log) putenv() failures, even
	 * if session handling is not enabled: very unlikely, and non-critical.
	 * Also handle cgroup; failures here are non-fatal */
	if(isopen){
		char *cp;

		/* XDG_RUNTIME_DIR */
		cp = xbuf;
		memcpy(cp, "XDG_RUNTIME_DIR=" a_RUNTIME_DIR_OUTER "/" a_RUNTIME_DIR_BASE "/",
			sizeof("XDG_RUNTIME_DIR=") -1 + sizeof(a_RUNTIME_DIR_OUTER) -1 + 1 +
				sizeof(a_RUNTIME_DIR_BASE) -1 + 1);
		cp += sizeof("XDG_RUNTIME_DIR=") -1 + sizeof(a_RUNTIME_DIR_OUTER) -1 + 1 +
				sizeof(a_RUNTIME_DIR_BASE) -1 + 1;
		memcpy(cp, &uidbuf[4], uidbuflen);

		if((res = pam_putenv(pamh, xbuf)) != PAM_SUCCESS)
			a_LOG(pamh, a_LOG_ERR, a_XDG ": user %s: pam_putenv(): %s\n", user, pam_strerror(pamh, res));

		/* {{{ And the rest unless runtime disallowed */
		if(!(f & a_RUNTIME)){
			struct a_dir{
				char const *name;
				size_t len;
				char const *defval;
			};

			static struct a_dir const a_dirs[] = {
				{"XDG_CACHE_HOME=", sizeof("XDG_CACHE_HOME=") -1, a_XDG_CACHE_HOME_DEF},
				{"XDG_CONFIG_DIRS=", sizeof("XDG_CONFIG_DIRS=") -1, a_XDG_CONFIG_DIRS_DEF},
				{"XDG_CONFIG_HOME=", sizeof("XDG_CONFIG_HOME=") -1, a_XDG_CONFIG_HOME_DEF},
				{"XDG_DATA_DIRS=", sizeof("XDG_DATA_DIRS=") -1, a_XDG_DATA_DIRS_DEF},
				{"XDG_DATA_HOME=", sizeof("XDG_DATA_HOME=") -1, a_XDG_DATA_HOME_DEF},
				{"XDG_STATE_HOME=", sizeof("XDG_STATE_HOME=") -1, a_XDG_STATE_HOME_DEF},
				{NULL,0,NULL} /* XXX -> nelem/item/countof */
			};

			char const *src;
			struct a_dir const *adp;
			size_t i;

			i = strlen(pwp->pw_dir);

			for(adp = a_dirs; adp->name != NULL; ++adp){
				cp = xbuf;
				memcpy(cp, adp->name, adp->len);
				cp += adp->len;
				if(*(src = adp->defval) == '\1'){
					memcpy(cp, pwp->pw_dir, i);
					cp += i;
					++src;
				}
				memcpy(cp, src, strlen(src) +1);

				if((res = pam_putenv(pamh, xbuf)) != PAM_SUCCESS)
					a_LOG(pamh, a_LOG_ERR, a_XDG ": user %s: pam_putenv(): %s\n",
						user, pam_strerror(pamh, res));
			}
		} /* }}} */

#if a_CGROUP /* {{{ */
		if(f & a_USER_CGROUP){
			int mode, e;
			size_t i;

			i = (size_t)(++cgroup_end - cgroup);
			memcpy(xbuf, cgroup, i);
			xbuf[i++] = '/';
			memcpy(&xbuf[i], &uidbuf[4], uidbuflen);

			if((res = openat(AT_FDCWD, xbuf, (a_O_SEARCH | O_DIRECTORY | O_NOFOLLOW))) != -1)
				close(res);
			/* See above for logic */
			else for(mode = 0700;;){
				gid_t oegid;
				mode_t oumask;

				oumask = umask(0000);
				oegid = getegid();
				setegid(0);
					res = mkdirat(AT_FDCWD, xbuf, mode);
					e = (res == -1) ? errno : 0;
				setegid(oegid);
				umask(oumask);

				if(res != -1){
					if(xbuf[i - 1] != '\0'){
						/* Simply chown it! */
						fchownat(AT_FDCWD, xbuf, pwp->pw_uid, pwp->pw_gid, AT_SYMLINK_NOFOLLOW);
						break;
					}
					xbuf[i - 1] = '/';
					mode = 0700;
				}else{
					if(xbuf[i - 1] == '\0')
						goto jcgroup_err;
					if(e != ENOENT){
jcgroup_err:
						a_LOG(pamh, a_LOG_ERR,
							a_XDG ": user %s: per_user_cgroup= unhandled: %s\n",
							user, strerror(e));
						goto jcgroup_done;
					}

					xbuf[i - 1] = '\0';
					mode = a_CGROUP_DIR_BASE_MODE;
				}
			}

			/* And write our PID */
			i += uidbuflen - 1;
			xbuf[i++] = '/';
			memcpy(&xbuf[i], a_CGROUP_PROCS_FILE, sizeof(a_CGROUP_PROCS_FILE));

			e = 0;
			if((res = openat(AT_FDCWD, xbuf, (O_WRONLY | O_NOFOLLOW | O_NOCTTY))) == -1 ||
					(mode = snprintf(pidbuf, sizeof pidbuf, "%lld\n", (long long int)getpid())
							) < 0 || write(res, pidbuf, (size_t)mode) == -1)
				e = errno;
			if(res != -1)
				close(res);
			if(e != 0)
				goto jcgroup_err;
jcgroup_done:;
		}
#endif /* a_CGROUP }}} */
	} /* }}} */

	/* In session mode we have to manage the counter file {{{ */
	if(f & a_SESSIONS){
		unsigned long long int sessions;

		/* Landed in the runtime base dir, obtain our lock */
		if((cntrlfd = openat(cwdfd, (f & a_USER_LOCK ? uidbuf : a_LOCK_FILE),
				(O_CREAT | O_WRONLY | O_NOFOLLOW | O_NOCTTY), (S_IRUSR | S_IWUSR))) == -1){
			emsg = "cannot open control lock file";
			goto jerr;
		}

		for(res = a_LOCK_TRIES;;){
			struct flock flp;

			memset(&flp, 0, sizeof flp);
			flp.l_type = F_WRLCK;
			flp.l_start = 0;
			flp.l_whence = SEEK_SET;
			flp.l_len = 0;

			if(fcntl(cntrlfd, F_SETLKW, &flp) != -1)
				break;

			/* XXX It may happen we cannot manage the lock and thus not access
			 * XXX the session counter, ie this session is zombie to us.
			 * XXX Just like counter below, should globally disable sessions!! */
			if(errno != EINTR){
				emsg = "unexpected error obtaining lock on lock control file";
				goto jerr;
			}
			if(--res == 0){
				emsg = "cannot obtain lock on lock control file";
				goto jerr;
			}
		}

		sessions = 0;

		if((datfd = openat(cwdfd, a_DAT_FILE, O_RDONLY)) != -1){
			char *ep;
			ssize_t r;

			while((r = read(datfd, xbuf, sizeof(xbuf) -1)) == -1){
				if(errno != EINTR)
					goto jecnt;
			}

			close(datfd);
			datfd = -1;

			xbuf[(size_t)r] = '\0';
			sessions = strtoull(xbuf, &ep, 10);
			/* Do not log too often for "session counter error"s, as below */
			if(ep == xbuf){
				/* It likely had been actively truncate(2)d below.  If we are opening this session,
				 * simply continue without session support, the corresponding log entry had been
				 * emitted in the past, so just give the user the environment variables (s)he wants */
				if(isopen){
					f |= a_SKIP_XDG;
					goto jok;
				}
				res = PAM_SESSION_ERR;
				goto jleave;
			}else if(sessions == ULLONG_MAX || ep != &xbuf[(size_t)r])
				goto jecnt;
		}

		if(isopen)
			++sessions;
		else if(sessions > 0)
			--sessions;

		if(/*!isopen &&*/ sessions == 0){
			/* Ridiculously simple, but everything else would be the opposite.
			 * Ie, E[MN]FILE failures, ordering issues, mode changes, mounts, subvolumes..
			 * This used to use system(3) because session opener programs were expected not to fiddle
			 * with $PATH in a way that would make this exploitable, but was then changed */
			static char const path[] = a_RUNTIME_DIR_OUTER "/" a_RUNTIME_DIR_BASE "/",
				rm[] = "/usr/bin/rm", *xargv[4] = {NULL, "-rf", NULL, NULL};
			pid_t xpid;

			if(access((emsg = &rm[sizeof("/usr") -1]), X_OK) && access((emsg = &rm[0]), X_OK)){
				emsg = "(/usr)?/bin/rm inaccessible, unable to rm(1) -rf per user XDG_RUNTIME_DIR";
				errno = EINVAL;
				goto jerr;
			}
			xargv[0] = emsg;

			memcpy(&xbuf[0], path, sizeof(path) -1);
			memcpy(&xbuf[sizeof(path) -1], &uidbuf[4], uidbuflen +1);
			xargv[2] = &xbuf[0];

			xpid = fork();
			if(xpid == -1)
				goto jeexec;
			if(xpid == 0){
				extern char **environ;

				execve(xargv[0], (char *const*)xargv, environ);
				exit(1);
			}

			waitpid(xpid, &res, 0);
			if(!WIFEXITED(res) || WEXITSTATUS(res) != 0){
jeexec:
				emsg = "unable to rm(1) -rf per user XDG_RUNTIME_DIR";
				errno = EINVAL;
				goto jerr;
			}
			/* This is the end .. */
			goto jok;
		}else{
			/* Write out session counter */
			res = snprintf(xbuf, sizeof xbuf, "%llu", sessions); /* xxx error? */

			if(((datfd = openat(cwdfd, a_DAT_FILE,
						(O_CREAT | O_TRUNC | O_WRONLY | O_SYNC | O_NOFOLLOW | O_NOCTTY),
						(S_IRUSR | S_IWUSR))) == -1) ||
					write(datfd, xbuf, res) != res){
jecnt:
				/* Ensure read above fails, so that henceforth session teardown is skipped for user */
				truncate(a_DAT_FILE, 0); /* xxx that may fail, too!? */
				emsg = "counter file error, disabled session tracking for user";
				goto jerr;
			}
			close(datfd);
			datfd = -1;
		}
	} /* }}} */

jok:
	res = PAM_SUCCESS;
jleave:
	if(datfd != -1)
		close(datfd);
	if(cntrlfd != -1)
		close(cntrlfd);
	if(cwdfd != -1 && cwdfd != AT_FDCWD) /* >=0, but AT_FDCWD unspecified */
		close(cwdfd);

	f &= a__SAVED_MASK;
	f_saved = f;

	return (res == PAM_SUCCESS) ? PAM_SUCCESS : PAM_SESSION_ERR;

jerr:
	a_LOG(pamh, a_LOG_ERR, a_XDG ": user %s: %s: %s\n", user, emsg, strerror(errno));
	f |= a_SKIP_XDG;
	res = PAM_SESSION_ERR;
	goto jleave;

jepam:
	a_LOG(pamh, a_LOG_ERR, a_XDG ": user %s: PAM failure: %s\n", user, pam_strerror(pamh, res));
	f |= a_SKIP_XDG;
	goto jleave;
}

int
pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv){
	return a_xdg(1, pamh, flags, argc, argv);
}

int
pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char **argv){
	return a_xdg(0, pamh, flags, argc, argv);
}

int
pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv){
	(void)flags;
	(void)argc;
	(void)argv;

	a_LOG(pamh, a_LOG_NOTICE, a_XDG ": pam_sm_acct_mgmt not used");

	return PAM_SERVICE_ERR;
}

int
pam_sm_setcred(pam_handle_t *pamh, int flags, int argc, const char **argv){
	(void)flags;
	(void)argc;
	(void)argv;

	a_LOG(pamh, a_LOG_NOTICE, a_XDG ": pam_sm_setcred not used");

	return PAM_SERVICE_ERR;
}

int
pam_sm_chauthtok(pam_handle_t *pamh, int flags, int argc, const char **argv){
	(void)flags;
	(void)argc;
	(void)argv;

	a_LOG(pamh, a_LOG_NOTICE, a_XDG ": pam_sm_chauthtok not used");

	return PAM_SERVICE_ERR;
}

/* s-itt-mode */
